### PR TITLE
HYACINTH-817: Don't allow disabling of fields that are in use

### PR DESCRIPTION
**Pull request by elohanlon**
- Also HYACINTH-915: EnabledDynamicFieldForm useReducer	refactor
- Also HYACINTH-916: EnabledDynamicFieldsForm should be able to select field sets for enabled fields

### CODE REVIEW

File: app/graphql/inputs/enabled_dynamic_field_input.rb [hunk 1]
```diff
@@ -7,9 +7,9 @@ class EnabledDynamicFieldInput < Types::BaseInputObject
     argument :dynamic_field, Inputs::IdReference, required: true
     argument :field_sets, [Inputs::IdReference], required: true
     argument :required, Boolean, required: true
-    argument :locked, Boolean, required: true
-    argument :hidden, Boolean, required: true
-    argument :owner_only, Boolean, required: true
+    argument :locked, Boolean, required: false # TODO: Get rid of this field if it's not useful
+    argument :hidden, Boolean, required: false # TODO: Get rid of this field if it's not useful
+    argument :owner_only, Boolean, required: false # TODO: Get rid of this field if it's not useful
```
[the remark this answers is not in the document]
Comment (barmintor) [reply app/graphql/inputs/enabled_dynamic_field_input.rb:12]: I like to have tickets to associate with TODOs - I created HYACINTH-923 for this, if you don't mind updating the comments

File: app/javascript/src/hyacinth_v1/components/projects/enabled_dynamic_fields/EnabledDynamicFieldShow.js [hunk 2]
```diff
@@ -38,11 +38,11 @@ export default class EnabledDynamicFieldShow extends React.Component {
             <EditButton
               className="float-end"
               size="lg"
+              aria-label="Edit"
```
Comment (barmintor) [inline app/javascript/src/hyacinth_v1/components/projects/enabled_dynamic_fields/EnabledDynamicFieldShow.js:41]: yes!

File: app/javascript/src/hyacinth_v1/components/projects/enabled_dynamic_fields/enabled_dynamic_field_form/EnabledDynamicFieldForm.js [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+      Object.values(searchTarget).forEach((val) => {
+        findDynamicFields(val, dynamicFields);
+      });
+    }
+  }
+  return dynamicFields;
+};
+
+const edfDefaultState = { enabledFieldHierarchy: [] };
+const edfReducer = (state, action) => {
```
Comment (barmintor) [inline app/javascript/src/hyacinth_v1/components/projects/enabled_dynamic_fields/enabled_dynamic_field_form/EnabledDynamicFieldForm.js:32]: if this is truly a function (that is, it does not refer to names in current scope), might want to consider a clean-up ticket to extract into a module.

File: lib/hyacinth/adapters/digital_object_search_adapter/solr.rb [hunk 4]
```diff
@@ -102,6 +78,42 @@ def identifier_to_uids(identifier, opts = {})
         def clear_index
           solr.clear
         end
+
```
Comment (barmintor) [inline lib/hyacinth/adapters/digital_object_search_adapter/solr.rb:81]: I would prefer to handle this via YARD documentation than actual access modifier - I don't think there's a concern about unintentional API use, and it makes testing annoying.

File: lib/hyacinth/digital_object/solr_keys.rb [hunk 5]
```diff
@@ -7,13 +7,18 @@ module SolrKeys
...
+        raise ArgumentError, "Path must be an array. Got: #{path.inspect}" unless path.is_a?(Array)
+        prefix = path.map { |p| p.camelcase(:lower) }.join('_')
         "#{prefix}_#{suffix}"
       end
 
       def self.for_dynamic_field(path, suffix = 'ssim')
         "df_#{for_string_key_path(path, suffix)}"
       end
+
+      def self.for_dynamic_field_presence(path)
```
Comment (barmintor) [inline lib/hyacinth/digital_object/solr_keys.rb:19]: this is a good idea - I'm tempted to ticket a similar change for language

File: lib/solr/utils.rb [hunk 6]
```diff
@@ -15,8 +15,14 @@ def self.suffix(data_type)
     # Escape characters that have special meaning to Solr query parser:
     # These include: + - & | ! ( ) { } [ ] ^ " ~ * ? : \ /
     # Does not escape whitespace. You need to do that on your own.
-    def self.escape(str)
-      RSolr.solr_escape(str)
+    def self.escape(val, escape_spaces = false)
```
Comment (barmintor) [inline lib/solr/utils.rb:18]: TY for parameterizing this!

File: spec/features/ui/v1/digital_objects/metadata/edit_spec.rb [hunk 7]
```diff
@@ -5,7 +5,7 @@
 RSpec.describe 'Digital Object Edit', solr: true, type: :feature, js: true do
   let(:uid) { SecureRandom.uuid }
   let(:item) { FactoryBot.create(:item, uid: uid) }
-  let(:permissons_required) { [] }
+  let(:permissions_required) { [] }
```
Comment (barmintor) [inline spec/features/ui/v1/digital_objects/metadata/edit_spec.rb:8]: don't be reluctant to pull this kind of change out and send it straight to `main` (if you don't mind the rebasing as you go)

File: lib/hyacinth/adapters/digital_object_search_adapter/solr.rb (hunk 4 above)
Comment (barmintor) [reply lib/hyacinth/adapters/digital_object_search_adapter/solr.rb:81]: Maybe even move the method to Solr::Utils?

**Review by barmintor — APPROVED** [review]
Good changes in here - my comments are all about documentation, and shouldn't block merge.

**Review by JackBlackLight — APPROVED** [review]

File: app/javascript/src/hyacinth_v1/components/projects/enabled_dynamic_fields/enabled_dynamic_field_form/EnabledDynamicFieldForm.js (hunk 3 above)
Comment (elohanlon) [reply app/javascript/src/hyacinth_v1/components/projects/enabled_dynamic_fields/enabled_dynamic_field_form/EnabledDynamicFieldForm.js]: Great suggestion!  Done for `edfReducer` and also `findDynamicFields`!

File: app/javascript/src/hyacinth_v1/components/projects/enabled_dynamic_fields/EnabledDynamicFieldShow.js (hunk 2 above)
Comment (elohanlon) [reply app/javascript/src/hyacinth_v1/components/projects/enabled_dynamic_fields/EnabledDynamicFieldShow.js:41]: 🎉

File: app/graphql/inputs/enabled_dynamic_field_input.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (elohanlon) [reply app/graphql/inputs/enabled_dynamic_field_input.rb]: I don't mind at all!  Thank you for creating the ticket.  Updated!

File: lib/solr/utils.rb (hunk 6 above)
Comment (elohanlon) [reply lib/solr/utils.rb:18]: 👍

File: lib/hyacinth/adapters/digital_object_search_adapter/solr.rb (hunk 4 above)
Comment (elohanlon) [reply lib/hyacinth/adapters/digital_object_search_adapter/solr.rb:85]: I think of Solr::Utils as having more generic solr functions -- ones that could be used in DigitalObjectSearchAdapter::Solr or TermSearchAdapter::Solr -- and these methods feel like they're tied more to DigitalObjectSearchAdapter, but I'm glad to remove the private designation and use YARD documentation instead.  I'll push an update for that right now.  Good suggestion.  👍

File: spec/features/ui/v1/digital_objects/metadata/edit_spec.rb (hunk 7 above)
Comment (elohanlon) [reply spec/features/ui/v1/digital_objects/metadata/edit_spec.rb]: Done!  Good idea, thanks!

**Comment by elohanlon** [comment]
Thank you both for reviewing and approving!  I've addressed all feedback and I'll merge as soon as CI finishes running.

**Pull request merged**